services/archiver/archiver.go

```diff
@@ -0,0 +1,306 @@
+	// Now we copy it into place
+	if destArchive, err = os.Create(r.archivePath); err != nil {
+		log.Error("Unable to open archive " + r.archivePath)
+		return
+	}
+	_, err = io.Copy(destArchive, tmpArchive)
+	destArchive.Close()
+	if err != nil {
+		log.Error("Unable to write archive " + r.archivePath)
+		return
+	}
+
+	// Block any attempt to finalize creating a new request if we're marking
+	r.archiveComplete = true
+}
+
+// ArchiveRepository satisfies the ArchiveRequest being passed in.  Processing
+// will occur in a separate goroutine, as this phase may take a while to
+// complete.  If the archive already exists, ArchiveRepository will not do
+// anything.  In all cases, the caller should be examining the *ArchiveRequest
+// being returned for completion, as it may be different than the one they passed
+// in.
+func ArchiveRepository(request *ArchiveRequest) *ArchiveRequest {
+	// We'll return the request that's already been enqueued if it has been
+	// enqueued, or we'll immediately enqueue it if it has not been enqueued
+	// and it is not marked complete.
+	archiveMutex.Lock()
+	defer archiveMutex.Unlock()
+	if rExisting := getArchiveRequest(request.repo, request.commit, request.archiveType); rExisting != nil {
+		return rExisting
+	}
+	if request.archiveComplete {
+		return request
+	}
+
+	request.cchan = make(chan struct{})
+	archiveInProgress = append(archiveInProgress, request)
+	go func() {
+		// Wait to start, if we have the Cond for it.  This is currently only
+		// useful for testing, so that the start and release of queued entries
+		// can be controlled to examine the queue.
+		if archiveQueueStartCond != nil {
+			archiveQueueMutex.Lock()
+			archiveQueueStartCond.Wait()
+			archiveQueueMutex.Unlock()
+		}
+
+		// Drop the mutex while we process the request.  This may take a long
+		// time, and it's not necessary now that we've added the reequest to
+		// archiveInProgress.
+		doArchive(request)
+
+		if archiveQueueReleaseCond != nil {
+			archiveQueueMutex.Lock()
+			archiveQueueReleaseCond.Wait()
+			archiveQueueMutex.Unlock()
+		}
+
+		// Purge this request from the list.  To do so, we'll just take the
+		// index at which we ended up at and swap the final element into that
+		// position, then chop off the now-redundant final element.  The slice
+		// may have change in between these two segments and we may have moved,
+		// so we search for it here.  We could perhaps avoid this search
+		// entirely if len(archiveInProgress) == 1, but we should verify
+		// correctness.
+		archiveMutex.Lock()
+		defer archiveMutex.Unlock()
+
+		idx := -1
+		for _idx, req := range archiveInProgress {
+			if req == request {
+				idx = _idx
+				break
+			}
+		}
+		if idx == -1 {
+			log.Error("ArchiveRepository: Failed to find request for removal.")
+			return
+		}
+		lastidx := len(archiveInProgress) - 1
+		if idx != lastidx {
+			archiveInProgress[idx] = archiveInProgress[lastidx]
+		}
+		archiveInProgress = archiveInProgress[:lastidx]
```

Comment:
Instead of swapping, you could just:
```
archiveInProgress = append(archiveInProgress[:idx], archiveInProgress[idx+1:]...);
```

https://play.golang.org/p/w1-d3aaCola

Reply:
So, I wasn't entirely sure here if I should be concerned about potential performance implications from slicing twice around the element, vs. the trivial swap to the end and cut it off.

Reply:
Slices in go are not arrays; this kind of manipulation is resolved internally using chunks and pointers (that's why they've got a name different from `array`). Super-performant.

Reply:
But we're also expecting a very short list of elements here.

Reply:
Right. ideally this queue should drain pretty quickly, but even for bots crawling somewhat large-ish installations... that's still pretty low order of magnitude. Readability is a little more important here, so I've switched it up.